### PR TITLE
Make `app deploy` to wait till deployment finishes 

### PR DESCRIPTION
**Pull request by zaiste**
Also enabled the `CI` environment variable along with the `--dispatch` option to skip the deployment await.

Improved and refined the command output.

CleanShot 2022-07-28 at 14 26 54@2x

### CODE REVIEW

**Comment by lkostrowski** [comment]
First look at these box-sections - awesome!!

File: src/cli/app/deploy.ts [hunk 1]
```diff
@@ -7,28 +8,37 @@ import got from 'got';
...
 import { Vercel } from '../../lib/vercel.js';
 import { useVercel } from '../../middleware/index.js';
+import { Options } from '../../types.js';
 
 export const command = 'deploy';
 export const desc = 'Deploy this Saleor App repository to Vercel';
 
-export const builder: CommandBuilder = (_) => _;
+export const builder: CommandBuilder = (_) =>
+  _.option('dispatch', {
```
Comment (lkostrowski) [inline src/cli/app/deploy.ts:23]: Maybe I'm not familiar with "dispatch" term in this context, so without reading description, I wouldn't firgure out what dispatch is. 
Is this term some pattern used in other apps? Will it be understandable for community?

File: src/cli/app/deploy.ts [hunk 2]
```diff
@@ -73,26 +83,39 @@ export const handler = async () => {
...
     projectId,
     newProject
   );
 
-  console.log(
-    '\nNext step - Install this Saleor App in your Saleor Dashboard\n'
-  );
-  console.log(`1. using Saleor Dashboard UI
-${chalk.cyan(
+  if (!!process.env.CI || !argv.dispatch) {
```
Comment (lkostrowski) [inline src/cli/app/deploy.ts:94]: I think it's good idea to have both ways. In many environments CI will be inferred by default and users will not have to remember about `dispatch` 👍

Comment (lkostrowski) [reply src/cli/app/deploy.ts:94]: Nitpick/suggestion - maybe I'd extract some util like `shouldWaitUntilDeployed = () => (!!process.env.CI || !argv.dispatch)` to add some semantics

File: src/cli/app/deploy.ts [hunk 3]
```diff
@@ -311,17 +338,30 @@ export const triggerDeploymentInVercel = async (
...
+      boxen(`${msg1} \n${msg2}`, {
+        padding: 1,
+        margin: 0,
+        borderColor: 'blue',
+        borderStyle: 'round',
+        title: 'Deployment URLs',
+      })
     );
-    console.log(`\nYour Vercel URL: ${chalk.cyan(`https://${url}`)}`);
+    console.log('');
```
Comment (lkostrowski) [inline src/cli/app/deploy.ts:362]: Is it needed?

File: src/lib/vercel.ts [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
   value: string;
   target?: string[];
   type?: string;
 }
 
+// interface Deployment {
+//   id: string
+//   url: string
+// }
+
```
Comment (lkostrowski) [inline src/lib/vercel.ts:17]: Probably we shouldn't keep dead code?

File: src/lib/vercel.ts [hunk 5]
```diff
@@ -110,4 +120,46 @@ export class Vercel {
   async getDeployments(projectId: string) {
     return this._client('GET', `/v6/deployments?projectId=${projectId}`);
   }
+
+  async getDeployment(deploymentId: string): Promise<any> {
```
Comment (lkostrowski) [inline src/lib/vercel.ts:124]: `Promise<any>` isn't type safe. If you perform runtime-check what is returned (which is good), it should be `Promise<unknown>`

File: src/lib/vercel.ts [hunk 6]
```diff
@@ -110,4 +120,46 @@ export class Vercel {
...
+    let { readyState } = await this.getDeployment(deploymentId);
+
+    if (deploymentFailed(readyState)) {
+      process.exit(1);
+    }
+
+    while (deploymentSucceeded(readyState)) {
+      await delay(5000);
+      const { readyState: rs } = await this.getDeployment(deploymentId);
+      readyState = rs;
```
Comment (lkostrowski) [inline src/lib/vercel.ts:143]: can we avoid mutation?

**Review by lkostrowski — CHANGES_REQUESTED** [review]

File: src/cli/app/deploy.ts (hunk 1 above)
Comment (zaiste) [reply src/cli/app/deploy.ts:23]: We can change this. Intuitively I used a term from the queueing nomenclature i.e. putting a long-running job on an execution queue is usually called _dispatching_; in other words we don't wait for the result, the execution is started and run asynchronously.

File: src/cli/app/deploy.ts (hunk 3 above)
Comment (zaiste) [reply src/cli/app/deploy.ts:362]: It's easier to see newlines this way when composing messages. With the `\n` at the end it's a bit difficult to see. I don't have a strong preference here. It's just slightly more practical.

File: src/cli/app/deploy.ts (hunk 2 above)
Comment (zaiste) [reply src/cli/app/deploy.ts:94]: Yes, good idea

**Comment by zaiste** [comment]
@lkostrowski PR refined. WDYT?

File: src/cli/app/deploy.ts (hunk 1 above)
Comment (lkostrowski) [reply src/cli/app/deploy.ts:23]: That sounds good 👍 

If we have a "better name" later (maybe more known for frontend devs) we can always add alias

File: src/cli/app/deploy.ts (hunk 3 above)
Comment (lkostrowski) [reply src/cli/app/deploy.ts:363]: How about 

```
console.log('\n');
```

It's a nitpick but this way it's pretty obvious what it does. But up2u

**Review by lkostrowski — APPROVED** [review]
LGTM 👏

**Review by 2can — APPROVED** [review]

File: src/cli/app/deploy.ts (hunk 3 above)
Comment (zaiste) [reply src/cli/app/deploy.ts:363]: Nitpicks are always welcome. Here, that would be a **double** newline, but let's discuss it on Monday. We have some ideas for a global approach.

Comment (lkostrowski) [reply src/cli/app/deploy.ts:363]: Ach, ok, I'm not used to CLI too much, was using mainly Ink for that. Its fine

**Comment by michalina-graczyk** [comment]
image
@zaiste There is still an `undefined` in the Dashboard URL

**Pull request merged**